### PR TITLE
Add casadi recipe

### PR DESCRIPTION
**Pull request by pstjohn**
Adds a recipe for [casadi](https://github.com/casadi/casadi/wiki), addressing casadi/casadi#1868
Currently I'm only planning to make a recipe for Mac/Linux, windows builds will likely be tricky with the chain of IPOPT dependencies

Lets see how this builds on CI. To be useful we'll most likely need
- [x] mumps-seq (working) conda-forge/mumps-feedstock#8
- [x] IPOPT #1607

### CODE REVIEW

**Comment by pstjohn** [comment]
Removing blasfeo (cf. [user group](https://groups.google.com/forum/#!topic/casadi-users/wiMtAQt5sMY)). Still not sure about the explicit openblas dependency or -DWITH_LAPACK.

**Comment by pstjohn** [comment]
@jakirkham, any ideas why this might be failing? It builds correctly with python 2.7, but when it switches over python 3.4 it doesn't re-download cmake.

**Comment by pstjohn** [comment]
@jakirkham, any idea why I might be getting the following error on circleCI?

```
[ 93%] Linking CXX executable ../../../bin/rocket_single_shooting
cd /staged-recipes/build_artefacts/casadi_1485656079538/work/build/docs/examples/cplusplus && /staged-recipes/build_artefacts/casadi_1485656079538/_b_env_place/bin/cmake -E cmake_link_script CMakeFiles/rocket_single_shooting.dir/link.txt --verbose=1
/staged-recipes/build_artefacts/casadi_1485656079538/_b_env_place/bin/g++    -std=c++11 -fPIC -O3 -DNDEBUG   CMakeFiles/rocket_single_shooting.dir/rocket_single_shooting.cpp.o  -o ../../../bin/rocket_single_shooting -Wl,-rpath,/staged-recipes/build_artefacts/casadi_1485656079538/work/build/lib -rdynamic ../../../lib/libcasadi_integrator_cvodes.so.3.1 ../../../lib/libcasadi_integrator_idas.so.3.1 ../../../lib/libcasadi_nlpsol_ipopt.so.3.1 ../../../lib/libcasadi_integrator_rk.so.3.1 ../../../lib/libcasadi_nlpsol_scpgen.so.3.1 ../../../lib/libcasadi_sundials_common.so.3.1 ../../../lib/libcasadi_sundials.so.3.1 -lipopt -ldmumps -lmumps_common -lpord -lmpiseq -lesmumps -lscotch -lscotcherr -lmetis -lgfortran -lgcc_eh -lopenblas -lm ../../../lib/libcasadi.so.3.1 -ldl -lrt 
/opt/rh/devtoolset-2/root/usr/bin/ld: warning: libquadmath.so.0, needed by /staged-recipes/build_artefacts/casadi_1485656079538/_b_env_place/bin/../lib/gcc/x86_64-unknown-linux-gnu/4.8.5/../../../../lib/libgfortran.so, not found (try using -rpath or -rpath-link)
/staged-recipes/build_artefacts/casadi_1485656079538/_b_env_place/bin/../lib/gcc/x86_64-unknown-linux-gnu/4.8.5/../../../../lib/libgfortran.so: undefined reference to `acoshq@QUADMATH_1.0'
```

I've tried adding libgfortran as a dependency but it doesnt seem to help. I thought openblas was supposed to provided the libquadmath dependency?

**Comment by pstjohn** [comment]
:tada:, seems to be working now. Pinging @conda-forge/core whenever you have a moment

File: recipes/casadi/build.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  -DWITH_SELFCONTAINED=OFF \
+  -DWITH_WORHP=OFF \
+  -DWITH_SLICOT=ON \
+  -DWITH_PYTHON=ON \
+  -DWITH_LAPACK=ON \
+  -DWITH_BLASFEO=OFF \
+  -DWITH_JSON=ON \
+  -DCMAKE_INSTALL_PREFIX:PATH=${PREFIX}\
+  -DPYTHON_PREFIX=${SP_DIR} \
+  ..
```
Comment (ocefpaf) [inline recipes/casadi/build.sh]: You can use `$SRC_DIR` instead of `..`

File: recipes/casadi/build.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+make VERBOSE=1
+make install
```
Comment (ocefpaf) [inline recipes/casadi/build.sh]: Is there a `make check` we can run here?

File: recipes/casadi/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  skip: true  # [win]
+  features:
+    - blas_{{ variant }}
+
+requirements:
+  build:
+    - cmake
+    - python
+    - swig
+    - numpy >=1.10
```
Comment (ocefpaf) [inline recipes/casadi/meta.yaml]: Does this packages uses `numpy`'s ABI? If so this must be `numpy x.x` here and below.

File: recipes/casadi/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+{% set version = "3.1.1" %}
+{% set variant = "openblas" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  git_url: https://github.com/casadi/casadi.git
+  git_rev: {{ version }}
```
Comment (ocefpaf) [inline recipes/casadi/meta.yaml]: Please change this to a tarball like https://github.com/ocefpaf/libdap4-feedstock/blob/master/recipe/meta.yaml#L7-L10 for example.

Comment (ocefpaf) [inline recipes/casadi/meta.yaml]: I see some `numpy.get_include()` in the source code, so the answer is probably yes.

File: recipes/casadi/build.sh (hunk 2 above)
Comment (pstjohn) [inline recipes/casadi/build.sh]: sadly no, just checked

File: recipes/casadi/meta.yaml (hunk 3 above)
Comment (pstjohn) [inline recipes/casadi/meta.yaml]: Sounds good. I had this in previous attempts but dropped it when I realized the official binaries don't specify a numpy version. (Also was getting some errors with numpy 1.9 in python 3.6, probably earlier version of the build script).

checking it locally before merging ... this is gonna take forever on CI :smile:

**Comment by pstjohn** [comment]
Ha, never seen that before. Travis ran out of log space.
I'm going to assume that the import test is sufficient to diagnose the major build problems... The `alltests.py` unittests are rather unnecessarily verbose

**Comment by pstjohn** [comment]
Well now I'm not sure what the best option is. Travis quits with: The job exceeded the maximum time limit for jobs, and has been terminated.

**Comment by scopatz** [comment]
Usually, you have to drop supported versions. Probably skipping all but the latest numpy is enough. If it isn't then try dropping Python version too.

**Comment by pstjohn** [comment]
I had thought the traditional advice was to compile against the oldest numpy you could, since theoretically the ABI should be forward-compatible. Thats partly why I had previously reverted to `numpy >= 1.10` in the dependencies.

What would be the correct way to restrict versions?

**Comment by scopatz** [comment]
I don't think it matters much to conda-forge which version you use. It is more about your need. The correct way to remove a version is to use `build: skip  # [selector]` http://conda-forge.github.io/docs/meta.html#excluding-a-platform

**Comment by pstjohn** [comment]
I guess my question is: if I pinned numpy to 1.10, would conda refuse to install this package for an environment that had a more recent numpy installed? 

(Also, is there an easy way to see the current build matrix? 😄 )

**Comment by scopatz** [comment]
> I pinned numpy to 1.10, would conda refuse to install this package for an environment that had a more recent numpy installed?

yes

> (Also, is there an easy way to see the current build matrix? :smile: )

I think conda-build-all will print it out at the top of all of the CIs

**Comment by pstjohn** [comment]
Cool, thanks for the help! Looks like we're building for 3 python versions + 2 numpy versions (11 + 12). 

We got partially through the 6th and final build on travis last time, lets see if the parallel build slips it in under the time limit. If not, I'll drop numpy 1.11

**Comment by pstjohn** [comment]
Finally :). Pinging @ocefpaf, does everything look OK at this point?

File: recipes/casadi/build.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  -DWITH_PYTHON=ON \
+  -DWITH_LAPACK=ON \
+  -DWITH_BLASFEO=OFF \
+  -DWITH_JSON=ON \
+  -DCMAKE_INSTALL_PREFIX:PATH=${PREFIX}\
+  -DPYTHON_PREFIX=${SP_DIR} \
+  $SRC_DIR
+
+make VERBOSE=0 -j${CPU_COUNT}
+make install
```
Comment (ocefpaf) [inline recipes/casadi/build.sh]: Maybe I already asked this, but is there a `make check` we can run here?

File: recipes/casadi/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+source:
+  fn: {{ version }}.tar.gz
+  url: https://github.com/casadi/casadi/archive/{{ version }}.tar.gz
+  sha256: 335eab9270a06d18134dafe34b5a21fe26fe6926611d7f960d826d82e851abef
+
+build:
+  number: 0
+  skip: true  # [win]
+  skip: true  # [np==111]
```
Comment (ocefpaf) [inline recipes/casadi/meta.yaml]: Just curious... Why are you skipping `np111`?

File: recipes/casadi/build.sh (hunk 5 above)
Comment (pstjohn) [inline recipes/casadi/build.sh]: i tried running one but got a `no rule to make target: check` or similar error

File: recipes/casadi/meta.yaml (hunk 6 above)
Comment (pstjohn) [inline recipes/casadi/meta.yaml]: as above, we run into errors on travis when we try to build this 6 times... (running over the 1h30 build time limit). So following @scopatz's suggestion, we skip all but the most recent numpy.

**Pull request merged**